### ISSUE TITLE
Question: `depends_on` and sub jobs making other jobs

### ISSUE THREAD

**Issue by andrewvaughanj** [rq/rq#1419]
## tl;dr

Is it possible to dynamically append to the dependencies of a job that hasn't started?

## Situation

In a workflow I am trying to create, I would like to have one job create an unknown number of children jobs, which in turn will create a further unknown number of grand-children jobs, and have a final job at the end that "joins" all the data at the end.

I guess conceptually it might look like this:

```
.
├── First job
│   └── n child jobs
│       └── m grand child jobs
└── Finalisation job
```

**Just to be clear**: I only want the finalisation job to run after all of the `m` grand-children have been processed; at the time of scheduling the first job and the finalisation job, *I have no idea how many children/grand-children will get created*.

I'm not building a web-crawler, but you can think of it like that:

```
# psuedo code
for website in websites:
     for page in website.index:
           download(page, output_dir)
process_all_pages(output_dir)
```

(but at the time of creating the `process_all_pages` job, I don't know how many websites are in `websites`)

## Code

This is some code that shows what I'd like to do. The `depends_on` line for the finalisation job is clearly not correct, because that job can run immediately after the first `discover_children` job has been created, and doesn't wait for the `process` to be called via `discover_grand_chidren`.

### `example.py`:

```python
#!/usr/bin/env python3

from rq import Connection, Queue

import example_lib


def main():
    q = Queue()

    discover_id = example_lib.fresh_uuid()
    join_id = example_lib.fresh_uuid()

    q.enqueue(example_lib.discover_children, job_id=discover_id)
    q.enqueue(example_lib.finalise, job_id=join_id, depends_on=[discover_id])


if __name__ == "__main__":
    with Connection():
        main()

# EOF

```

### `example_lib.py`:

```python
from rq import Connection, Queue
import uuid
import os

FNAME = "/tmp/moo"
N = 10
M = 10


def fresh_uuid():
    return str(uuid.uuid4())


def discover_children():
    print("discover_children")
    q = Queue()

    for elem in range(N):
        q.enqueue(discover_grand_chidren, elem)


def discover_grand_chidren(child_idx):
    print("discover_grand_chidren")
    q = Queue()

    for elem in range(M):
        q.enqueue(process, child_idx, elem)


def process(child_idx, grand_child_idx):
    if child_idx == N - 1 and grand_child_idx == M - 1:
        open(FNAME, "w").close()


def finalise():
    print("finalise")
    assert os.path.exists(FNAME)


# EOF
```

### Idea 1

One idea I had was if I could pass the `job_id` for the finalisation job to all of the children/grandchildren, and when creating the child/grandchild jobs, "append" to the list of dependencies for the finalisation job.

However, while looking in the `rq` code, it seems that we have no way to "dynamically" reload the dependencies, so I cannot append to that list.

### Idea 2

Another possibility was to create a "final" job at each level (so one for the grandchildren and one for the parent), that can get rescheduled if the children aren't all done.

Something like:

```
for child in children:
      if child.still_processing:
             raise Reschedule
```

but there doesn't seem to be anything like this in the `rq` code, and I am not sure how I can "abuse" Redis to store if all the children are done or not.

### Idea 3 (which won't work)

Add a `while not done` loop to each level of jobs, that then "blocks" (so the child job never finishes until the grandchildren are done).

The problem with this that it will sit on a worker while waiting for the children to finish. I'd really like to avoid this, as it will needlessly consume workers that could be doing real work.

**Comment by andrewvaughanj** [comment]
Okay, so I have a "working solution", but I don't consider it particularly graceful:

### `example.py`

```python
#!/usr/bin/env python3

from rq import Connection, Queue, Retry

import example_lib

# Give me a huge recount number
huge_int = ((9 ** 9) ** 9) ** 9


def main():
    q = Queue()

    discover_id = example_lib.fresh_uuid()
    join_id = example_lib.fresh_uuid()

    # Queue-up our discovery job
    q.enqueue(example_lib.discover, job_id=discover_id)

    # Queue-up our finalise job, but with a huge retry count!
    q.enqueue(
        example_lib.finalise,
        job_id=join_id,
        retry=Retry(max=huge_int),
        depends_on=[discover_id],
    )


if __name__ == "__main__":
    with Connection():
        main()

# EOF
```

### `example_lib.py`:

```python
(venv) avj@moss:~/clones/atg_redis/src> cat example_lib.py 
from rq import Connection, Queue
from redis import Redis
from rq.job import Job, JobStatus
from rq import get_current_job
import uuid
import os

FNAME = "/tmp/moo"
N = 10
M = 10


def fresh_uuid():
    return str(uuid.uuid4())


def discover(parent_id=None):
    q = Queue()

    all_children = []
    for elem in range(N):
        job_id = fresh_uuid()
        all_children.append(job_id)
        if parent_id is None:
            q.enqueue(discover, parent_id=elem, job_id=job_id)
        else:
            q.enqueue(process, parent_id, elem, job_id=job_id)

    # Record in the meta data for the current job what our newly created
    # children are
    job = get_current_job()
    job.meta["children"] = all_children
    job.save_meta()


def process(child_idx, grand_child_idx):
    if child_idx == (N - 1) and grand_child_idx == (M - 1):
        open(FNAME, "w").close()


def check_children_all_done():
    """
    Recursively check if all dependencies and all children are finished
    """

    # Connect to redis to get jobs
    redis = Redis()

    # Get the current job
    curr_job = get_current_job()
    curr_job_id = curr_job.get_id()

    #
    # We store the id to avoid fetching jobs unncessarily (this pays a penalty
    # on fetching the current job)
    #
    work_list = set([curr_job_id])

    # We're now going to process our current job + all it chldren
    while work_list:

        # Grab one job id from the work list
        one_job_id = work_list.pop()

        # Get this job
        one_job = Job.fetch(one_job_id, connection=redis)

        # If we are not the current job (current job will always be running!)
        if one_job_id != curr_job_id:

            # Get the status
            one_job_status = one_job.get_status()

            # Is this child finished?
            if one_job_status != JobStatus.FINISHED:

                #
                # IMPORTANT BIT:
                #
                # If one child isn't finished, we want to reschedule the
                # current job
                #
                # This throws *** which is perfect ***!!!
                #
                curr_job.requeue()

        # Get all of our explicit dependencies
        one_deps = one_job.dependency_ids

        # Iterate on them
        for one_dep in one_deps:

            # Need to parse the job id -- yuck!
            job_id = one_dep.decode("utf-8")
            child_job_id = job_id.split(":")[2]

            # Store that we want to process it
            work_list.add(child_job_id)

        # Do we have 'children' in our meta?
        if "children" in one_job.meta:

            # If we do, grab the children
            children = one_job.meta["children"]

            # For each explicit child
            for child_job_id in children:
                # Store that we want to process it
                work_list.add(child_job_id)

    # We only get here if all dependencies and all children are finished!
    return True


def finalise():
    """"""

    # This will always hold -- either we throw (as part of the call) or the
    # function return true
    assert check_children_all_done()

    assert os.path.exists(FNAME)
    print("ALL DONE")


# EOF
```

## Question

Is there a way to make a job have an "infinite" number of retries?

**Comment by andrewvaughanj** [comment]
> Is there a way to make a job have an "infinite" number of retries?
>

It looks like (and I am not 100% that this does work), but that you can change `curr_job.retries_left` _before_ doing the reschedule.

If you do this, you can keep setting the number of retries to 1, and then `rq` continually re-tries.

**Comment by selwin** [comment]
> However, while looking in the rq code, it seems that we have no way to "dynamically" reload the dependencies, so I cannot append to that list.

We can add an API to add a dependency to a job if that's what you need, so something like `job.add_dependency(job_2)`. But even with this, the solution won't be bullet proof because you have to consider the possibility of race conditions.

**Comment by andrewvaughanj** [comment]
Woah, that could be really great!

Are you saying "racy" in terms of making sure the job we want to append to hasn't yet started, or "racy" in terms of two jobs trying to add a dependency concurrently?

I actually don't think it is possible for my use-case that this could be racy for the first kind (appending after it has already started) -- these dependent jobs would only get appended by *already* dependent job. As long as the "finish" job does not start before any of its current dependents (which it shouldn't, be design), one dependent appending to the list of dependencies should be fine, as long as it can be done by many jobs concurrently.

There is a caveat here that I'd probably need to use a pipeline to create all the jobs "at the top level" to stop one child trying to append to the finish job, in case it hasn't yet been created.

Would it be possible for `add_dependency` to take a list of jobs?

**Comment by selwin** [comment]
> Are you saying "racy" in terms of making sure the job we want to append to hasn't yet started, or "racy" in terms of two jobs trying to add a dependency concurrently?

Obviously the implementation on RQ's end needs to handle race conditions, but you application logic could also have race conditions.

> Would it be possible for add_dependency to take a list of jobs?

It could, now we just need someone to work on it ;)

**Comment by andrewvaughanj** [comment]
@selwin I am 100% happy to give this ago -- I am contributor to a number of other open source projects, but I have a few questions as I am new to `rq` (as a contributor).

After looking at things, it seems that there are few main areas where dependencies are **read**:

- [`Job.dependencies_are_met`](https://github.com/rq/rq/blob/016da14723d21f6801b19cbff97f2fb9bee5c5db/rq/job.py#L819)
- [`Queue.enqueue_call`](https://github.com/rq/rq/blob/016da14723d21f6801b19cbff97f2fb9bee5c5db/rq/queue.py#L313)
- [`Queue.enqueue_dependents`](https://github.com/rq/rq/blob/016da14723d21f6801b19cbff97f2fb9bee5c5db/rq/queue.py#L488)
- [`Worker.handle_job_success`](https://github.com/rq/rq/blob/016da14723d21f6801b19cbff97f2fb9bee5c5db/rq/worker.py#L925)

Are we happy that all of these will be okay if we randomly change the value of `_dependency_ids` while any of these are executing?

Just to reinforce this, for my use-case, I will only be adding new dependencies in already dependent jobs (i.e., if I am adding to the dependencies of Job `foo`, while executing Job `bar`, then `foo` *already* depends on `bar`, so we know that `foo` won't be executing -- but I guess a worker might be running some of the methods above to check if all dependencies are satisfied).

If the above is fine, then would an appropriate solution be add a new method that allows us to *append* to `Job._dependency_ids` + then invoke [`Job.register_dependency`](https://github.com/rq/rq/blob/016da14723d21f6801b19cbff97f2fb9bee5c5db/rq/job.py#L788)?

**Comment by andrewvaughanj** [comment]
@selwin any thoughts on the best way to start this?

**Comment by selwin** [comment]
@andrewvaughanj sorry I missed replying to this thread. Implementing a `job.add_dependency()` that takes a list of jobs like your previous suggestion is a good way to start.

**Comment by andrewvaughanj** [comment]
Okay, I can take a look at this!

**Comment by wattux** [comment]
@aytey Any updates on this? Do you still plan to give a PR for this a try? I'd love to see this function being added, and could attempt a PR too.
